### PR TITLE
New scheduler local node

### PR DESCRIPTION
**Pull request by istoica**
## Why are these changes needed?
Added support for handling resources with multiple instances (e.g., GPU and CPU) to the new scheduler. Added tests. 

The last thing that remains to be done is to call the new API for handling resource instances from node_manager.cc and removing the old code.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

File: src/ray/common/scheduling/cluster_resource_scheduler.cc [hunk 1]
```diff
@@ -1,11 +1,42 @@
...
+      return false;
+    }
+    if (this->predefined_resources[i].available !=
+        other.predefined_resources[i].available) {
+      return false;
+    }
+  }
+
+  if (this->custom_resources.size() != other.custom_resources.size()) {
+    return true;
```
Comment (edoakes) [inline src/ray/common/scheduling/cluster_resource_scheduler.cc]: Shouldn't this be `return false`?

File: src/ray/common/scheduling/cluster_resource_scheduler.cc [hunk 2]
```diff
@@ -19,6 +50,87 @@ std::string NodeResources::DebugString() {
   return buffer.str();
 }
 
+std::string PrintVector(std::vector<double> &vector) {
```
Comment (edoakes) [inline src/ray/common/scheduling/cluster_resource_scheduler.cc]:
```suggestion
std::string DebugString(std::vector<double> &vector) {
```
Doesn't actually print, just returns a `std::string`

File: src/ray/common/scheduling/cluster_resource_scheduler.cc [hunk 3]
```diff
@@ -19,6 +50,87 @@ std::string NodeResources::DebugString() {
...
+      return false;
+    }
+    if (!EqualVectors(this->predefined_resources[i].available,
+                      other.predefined_resources[i].available)) {
+      return false;
+    }
+  }
+
+  if (this->custom_resources.size() != other.custom_resources.size()) {
+    return true;
```
Comment (edoakes) [inline src/ray/common/scheduling/cluster_resource_scheduler.cc]: Shouldn't this be `return false`?

File: src/ray/common/scheduling/cluster_resource_scheduler.cc [hunk 4]
```diff
@@ -19,6 +50,87 @@ std::string NodeResources::DebugString() {
...
+
+TaskResourceInstances NodeResourceInstances::GetAvailableResourceInstances() {
+  TaskResourceInstances task_resources;
+  task_resources.predefined_resources.resize(PredefinedResources_MAX);
+
+  for (size_t i = 0; i < this->predefined_resources.size(); i++) {
+    task_resources.predefined_resources[i] = this->predefined_resources[i].available;
+  }
+
+  for (auto it = this->custom_resources.begin(); it != this->custom_resources.end();
```
Comment (edoakes) [inline src/ray/common/scheduling/cluster_resource_scheduler.cc]: Prefer to use `for (const auto it : this->custom_resources) {`

File: src/ray/common/scheduling/cluster_resource_scheduler.h [hunk 5]
```diff
@@ -63,15 +60,59 @@ struct TaskRequest {
   std::string DebugString();
 };
 
+// Task request specifying instances for each resource.
+struct TaskResourceInstances {
```
Comment (edoakes) [inline src/ray/common/scheduling/cluster_resource_scheduler.h]: Borderline, but I'd prefer to make these classes

https://google.github.io/styleguide/cppguide.html#Structs_vs._Classes

File: src/ray/common/scheduling/cluster_resource_scheduler.h [hunk 6]
```diff
@@ -227,9 +265,83 @@ class ClusterResourceScheduler {
...
+
+  /// Create instances for each resource associated with the local node, given
+  /// the node's resources.
+  ///
+  /// \param local_resources: Total resources of the node.
+  void InitLocalResources(const NodeResources &local_resources);
+
+  /// Initialize the instances of a given resource given the resource's total capacity.
+  /// If unit_instances is true we split the resources in unit-size instances. For
+  /// example, if total = 10, then we create 10 instances, each with caoacity 1.
```
Comment (edoakes) [inline src/ray/common/scheduling/cluster_resource_scheduler.h]:
```suggestion
  /// example, if total = 10, then we create 10 instances, each with capacity 1.
```

File: src/ray/common/scheduling/cluster_resource_scheduler.h [hunk 7]
```diff
@@ -227,9 +265,83 @@ class ClusterResourceScheduler {
...
+  /// Otherwise, we create a single instance of capacity equal to the resource's capacity.
+  ///
+  /// \param total: Total resource capacity.
+  /// \param unit_instances: If true, we split the resource in unit-size instances.
+  /// If false, we create a single instance of capacity "total".
+  /// \param instance_list: The list of capacities this resource instances.
+  void InitResourceInstances(double total, bool unit_instances,
+                             ResourceInstanceCapacities *instance_list);
+
+  /// Allocate enough capacity across the instances of a resource to satisfy "demand".
```
Comment (edoakes) [inline src/ray/common/scheduling/cluster_resource_scheduler.h]:
```suggestion
  /// Allocate enough capacity across the instances of a resource to satisfy `demand`.
```

File: src/ray/common/scheduling/cluster_resource_scheduler.h [hunk 8]
```diff
@@ -227,9 +265,83 @@ class ClusterResourceScheduler {
...
+  ///
+  /// \param demand: The resource amount to be allocated.
+  /// \param soft: Specifies whether this demand has soft or hard constraints.
+  /// \param available: List of available capacities of the instances of the resource.
+  /// \param allocation: List of instance capacities allocated to satisfy the demand.
+  /// This is a return parameter.
```
Comment (edoakes) [inline src/ray/common/scheduling/cluster_resource_scheduler.h]: Please use `\param[in]` and `\param[out]` to annotate this

File: src/ray/common/scheduling/cluster_resource_scheduler.h [hunk 9]
```diff
@@ -227,9 +265,83 @@ class ClusterResourceScheduler {
...
+  /// Decrease the available capacities of the instances of a given resource.
+  ///
+  /// \param free A list of capacities for resource's instances to be freed.
+  /// \param resource_instances List of the resource instances being updated.
+  void SubtractAvailableResourceInstances(std::vector<double> free,
+                                          ResourceInstanceCapacities *resource_instances);
+
+  /// Increase the available cpu instances of this node. 
+  ///
+  /// \param cpu_instances Cpu instances to be added to available cpus.
```
Comment (edoakes) [inline src/ray/common/scheduling/cluster_resource_scheduler.h]:
```suggestion
  /// \param cpu_instances CPU instances to be added to available cpus.
```

File: src/ray/common/scheduling/cluster_resource_scheduler.h [hunk 10]
```diff
@@ -227,9 +265,83 @@ class ClusterResourceScheduler {
...
+  void AddCPUResourceInstances(std::vector<double> &cpu_instances);
+
+  /// Decrease the available cpu instances of this node. 
+  ///
+  /// \param cpu_instances Cpu instances to be removed from available cpus.
```
Comment (edoakes) [inline src/ray/common/scheduling/cluster_resource_scheduler.h]:
```suggestion
  /// \param cpu_instances CPU instances to be removed from available cpus.
```

File: src/ray/common/scheduling/cluster_resource_scheduler.h [hunk 11]
```diff
@@ -227,9 +265,83 @@ class ClusterResourceScheduler {
...
+  /// \param resource_instances List of the resource instances being updated.
+  void SubtractAvailableResourceInstances(std::vector<double> free,
+                                          ResourceInstanceCapacities *resource_instances);
+
+  /// Increase the available cpu instances of this node. 
+  ///
+  /// \param cpu_instances Cpu instances to be added to available cpus.
+  void AddCPUResourceInstances(std::vector<double> &cpu_instances);
+
+  /// Decrease the available cpu instances of this node.
```
Comment (edoakes) [inline src/ray/common/scheduling/cluster_resource_scheduler.h]:
```suggestion
  /// Decrease the available CPU instances of this node. 
```

File: src/ray/common/scheduling/cluster_resource_scheduler.h [hunk 12]
```diff
@@ -227,9 +265,83 @@ class ClusterResourceScheduler {
...
+                                     ResourceInstanceCapacities *resource_instances);
+
+  /// Decrease the available capacities of the instances of a given resource.
+  ///
+  /// \param free A list of capacities for resource's instances to be freed.
+  /// \param resource_instances List of the resource instances being updated.
+  void SubtractAvailableResourceInstances(std::vector<double> free,
+                                          ResourceInstanceCapacities *resource_instances);
+
+  /// Increase the available cpu instances of this node.
```
Comment (edoakes) [inline src/ray/common/scheduling/cluster_resource_scheduler.h]:
```suggestion
  /// Increase the available CPU instances of this node. 
```

File: src/ray/common/scheduling/cluster_resource_scheduler.cc [hunk 13]
```diff
@@ -487,3 +621,225 @@ std::string ClusterResourceScheduler::DebugString(void) {
...
+        (*allocation)[i] = available[i];
+        demand -= available[i];
+        available[i] = 0;
+      }
+    }
+    return true;
+  }
+
+  if (demand >= 1.) {
+    // Cannot satisfy a deman greater than one if no unit caapcity resource is available.
```
Comment (edoakes) [inline src/ray/common/scheduling/cluster_resource_scheduler.cc]: typos

File: src/ray/common/scheduling/cluster_resource_scheduler.cc [hunk 14]
```diff
@@ -487,3 +621,225 @@ std::string ClusterResourceScheduler::DebugString(void) {
...
+      if (soft) {
+        available[0] = 0;
+        return true;
+      }
+      // Not enough capacity.
+      return false;
+    }
+  }
+
+  // If resources has multiple instances, each instance has total capcity of 1.
```
Comment (edoakes) [inline src/ray/common/scheduling/cluster_resource_scheduler.cc]:
```suggestion
  // If resources has multiple instances, each instance has total capacity of 1.
```

File: src/ray/common/scheduling/cluster_resource_scheduler.cc [hunk 15]
```diff
@@ -487,3 +621,225 @@ std::string ClusterResourceScheduler::DebugString(void) {
...
+  // If demand is greater than 1., allocate full unit-capacity instances until the
+  // remaining demand is fractional.
```
Comment (edoakes) [inline src/ray/common/scheduling/cluster_resource_scheduler.cc]: extend comment to explain why we break once it's fractional/what happens afterward?

File: src/ray/common/scheduling/cluster_resource_scheduler.cc [hunk 16]
```diff
@@ -487,3 +621,225 @@ std::string ClusterResourceScheduler::DebugString(void) {
...
+        demand -= 1.;
+      }
+      if (demand < 1.) {
+        break;
+      }
+    }
+  }
+
+  if (soft) {
+    // Just get as much resources as available.
```
Comment (edoakes) [inline src/ray/common/scheduling/cluster_resource_scheduler.cc]:
```suggestion
    // Just get as many resources as available.
```

File: src/ray/common/scheduling/cluster_resource_scheduler.cc (hunk 13 above)
Comment (edoakes) [reply src/ray/common/scheduling/cluster_resource_scheduler.cc]: Seems like this error check should be higher up in the function, before we start modifying `demand`

Comment (edoakes) [reply src/ray/common/scheduling/cluster_resource_scheduler.cc]: Also, in general it's preferable to not modify the arguments to the function. Better to make a separate variable (e.g., `remaining_demand` so it's clear when you're checking against the passed argument vs. current state of the function.

File: src/ray/common/scheduling/cluster_resource_scheduler.cc (hunk 1 above)
Comment (istoica) [reply src/ray/common/scheduling/cluster_resource_scheduler.cc]: Yes, great find!

File: src/ray/common/scheduling/cluster_resource_scheduler.cc (hunk 2 above)
Comment (istoica) [reply src/ray/common/scheduling/cluster_resource_scheduler.cc]: Ok, I'll change it to DebugString

File: src/ray/common/scheduling/cluster_resource_scheduler.cc (hunk 3 above)
Comment (istoica) [reply src/ray/common/scheduling/cluster_resource_scheduler.cc]: Right again! (BTW, bot this and the previous similar mistake occurs in methods that have been only used for testing ;-)

File: src/ray/common/scheduling/cluster_resource_scheduler.h (hunk 5 above)
Comment (istoica) [reply src/ray/common/scheduling/cluster_resource_scheduler.h]: Let's discuss this in person, and maybe do it in the next PR. If classes, I'd still like to have the members public (not have another method accessing them) due to performance reasons. But then, why do a class? Again, let's talk in person.

File: src/ray/common/scheduling/cluster_resource_scheduler.cc (hunk 13 above)
Comment (istoica) [reply src/ray/common/scheduling/cluster_resource_scheduler.cc]: The intention that unless the resources are soft we only allocate at most one fractional resource. For instance. If we ask for 2.5 CPUs, we get two full CPUs and 0.5 of another CPU, not 0.25 of one CPU and another 0.25 of another CPU. I'll make that clear.

Regarding remaining_demand, that sounds good.

File: src/ray/common/scheduling/cluster_resource_scheduler.cc (hunk 4 above)
Comment (istoica) [reply src/ray/common/scheduling/cluster_resource_scheduler.cc]: fixed

File: src/ray/common/scheduling/cluster_resource_scheduler.cc (hunk 14 above)
Comment (istoica) [reply src/ray/common/scheduling/cluster_resource_scheduler.cc]: done

File: src/ray/common/scheduling/cluster_resource_scheduler.cc (hunk 2 above)
Comment (edoakes) [reply src/ray/common/scheduling/cluster_resource_scheduler.cc]: When I read `PrintVector`, I expected it to write the string to stderr/stdout, but this just returns a string so I'd prefer to call it something like `DebugString`

File: src/ray/common/scheduling/cluster_resource_scheduler.cc (hunk 13 above)
Comment (edoakes) [reply src/ray/common/scheduling/cluster_resource_scheduler.cc]: Are you planning to modify this to use `remaining_demand`?

File: src/ray/common/scheduling/cluster_resource_scheduler.cc (hunk 16 above)
Comment (edoakes) [reply src/ray/common/scheduling/cluster_resource_scheduler.cc]: ping

File: src/ray/common/scheduling/cluster_resource_scheduler.h (hunk 5 above)
Comment (edoakes) [reply src/ray/common/scheduling/cluster_resource_scheduler.h]: It's just a convention that makes it easier to understand the code without context - `struct`s should generally only be used to wrap data and not to define. This also shouldn't have any impact on performance - classes and structs are completely the same under the hood AFAIK and getters/setters will be optimized away.

File: src/ray/common/scheduling/cluster_resource_scheduler.h (hunk 6 above)
Comment (edoakes) [reply src/ray/common/scheduling/cluster_resource_scheduler.h]: ping

**Review by edoakes — COMMENTED** [review]
LGTM but I think you missed a few of my previous comments, once those are addressed can merge

File: src/ray/common/scheduling/cluster_resource_scheduler.cc (hunk 2 above)
Comment (istoica) [reply src/ray/common/scheduling/cluster_resource_scheduler.cc]: Ah ok. I changed it to `Vector2String`. Cannot call it `DebugString` as it is conflicting with the identical named method of other classes.

File: src/ray/common/scheduling/cluster_resource_scheduler.h (hunk 12 above)
Comment (istoica) [reply src/ray/common/scheduling/cluster_resource_scheduler.h]: done

File: src/ray/common/scheduling/cluster_resource_scheduler.cc (hunk 13 above)
Comment (istoica) [reply src/ray/common/scheduling/cluster_resource_scheduler.cc]: Already did it!

File: src/ray/common/scheduling/cluster_resource_scheduler.cc [hunk 17]
```diff
@@ -50,7 +50,7 @@ std::string NodeResources::DebugString() {
   return buffer.str();
 }
 
-std::string PrintVector(std::vector<double> &vector) {
+std::string Vector2String(std::vector<double> &vector) {
```
Comment (edoakes) [inline src/ray/common/scheduling/cluster_resource_scheduler.cc]:
```suggestion
std::string VectorToString(std::vector<double> &vector) {
```

**Review by edoakes — APPROVED** [review]
LGTM, just one small nit left

**Comment by edoakes** [comment]
Linter is also failing:
https://travis-ci.com/ray-project/ray/jobs/282873093

**Pull request merged**